File: C/src/dynamic_queue.c
#include <stdio.h>
#include <stdlib.h>
#include "dynamic_queue.h"

void inicializarDynamicQueue(DynamicQueue *queue) {
    queue->head = NULL;
    queue->tail = NULL;
    queue->size = 0;
}

int enfileirarDynamicQueue(DynamicQueue *queue, Registro valor) {
    No *novo = (No*) malloc(sizeof(No));
    if (!novo) return 0;

    novo->valor = valor;
    novo->prox = NULL;

    if (queue->tail == NULL) {
        queue->head = novo;
        queue->tail = novo;
    } else {
        queue->tail->prox = novo;
        queue->tail = novo;
    }

    queue->size++;
    return 1;
}

int desenfileirarDynamicQueue(DynamicQueue *queue, Registro *valor) {
    if (filaVazia(queue)) return 0;

    No *remover = queue->head;
    *valor = remover->valor;
    queue->head = remover->prox;
    if (queue->head == NULL)
        queue->tail = NULL;

    free(remover);
    queue->size--;
    return 1;
}

int filaVaziaDynamicQueue(DynamicQueue *queue) {
    return queue->size == 0;
}

void imprimirDynamicQueue(DynamicQueue *queue) {
    No *atual = queue->head;
    while (atual != NULL) {
        printf("%d ", atual->valor.valor);
        atual = atual->prox;
    }
    printf("\n");
}

void salvarDynamicQueueEmArquivo(DynamicQueue *queue, const char *nomeArquivo) {
    FILE *arquivo = fopen(nomeArquivo, "w");
    if (!arquivo) {
        perror("Erro ao abrir arquivo para escrita");
        return;
    }

    No *atual = queue->head;
    while (atual != NULL) {
        fprintf(arquivo, "%d\n", atual->valor.valor);
        atual = atual->prox;
    }

    fclose(arquivo);
}

File: Java/src/DynamicQueue.java
public class DynamicQueue {
    protected Node head;
    protected Node tail;
    protected int size;

    public DynamicQueue() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public boolean enfileirar(Registro valor) {
        Node novo = new Node(valor);

        if (this.tail == null) {
            this.head = novo;
            this.tail = novo;
        } else {
            this.tail.prox = novo;
            this.tail = novo;
        }

        this.size++;
        return true;
    }

    public Registro desenfileirar() {
        if (filaVazia()) return null;

        Node remover = this.head;
        Registro valor = remover.valor;
        this.head = remover.prox;
        if (this.head == null)
            this.tail = null;

        this.size--;
        return valor;
    }

    public boolean filaVazia() {
        return this.size == 0;
    }

    public void imprimir() {
        Node atual = this.head;
        while (atual != null) {
            System.out.print(atual.valor.valor + " ");
            atual = atual.prox;
        }
        System.out.println();
    }

    public void salvarEmArquivo(String nomeArquivo) {
        try (java.io.FileWriter writer = new java.io.FileWriter(nomeArquivo)) {
            Node atual = this.head;
            while (atual != null) {
                writer.write(atual.valor.valor + "\n");
                atual = atual.prox;
            }
        } catch (java.io.IOException e) {
            System.err.println("Erro ao abrir arquivo para escrita: " + e.getMessage());
        }
    }
}

